x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java

```diff
@@ -657,6 +659,95 @@ public void getUserPrivileges(AuthorizationInfo authorizationInfo, ActionListene
     }
 
+    @Override
+    public void getRemoteAccessRoleDescriptorsIntersection(
+        String remoteClusterAlias,
+        AuthorizationInfo authorizationInfo,
+        ActionListener<RoleDescriptorsIntersection> listener
+    ) {
+        if (authorizationInfo instanceof RBACAuthorizationInfo == false) {
+            listener.onFailure(
+                new IllegalArgumentException("unsupported authorization info:" + authorizationInfo.getClass().getSimpleName())
+            );
+            return;
+        }
+
+        final Role role = ((RBACAuthorizationInfo) authorizationInfo).getRole();
+        final RemoteIndicesPermission remoteIndicesPermission;
+        try {
+            // TODO we will need to implement this to support API keys with assigned role descriptors
+            remoteIndicesPermission = role.remoteIndices().forCluster(remoteClusterAlias);
+        } catch (UnsupportedOperationException e) {
+            listener.onFailure(
+                new IllegalArgumentException(
+                    "Cannot retrieve remote access role descriptors for API keys with assigned role descriptors.",
+                    e
+                )
+            );
+            return;
+        }
+        if (remoteIndicesPermission.remoteIndicesGroups().isEmpty()) {
+            listener.onResponse(RoleDescriptorsIntersection.EMPTY);
+            return;
+        }
+
+        final List<RoleDescriptor.IndicesPrivileges> indicesPrivileges = new ArrayList<>();
+        for (RemoteIndicesPermission.RemoteIndicesGroup remoteIndicesGroup : remoteIndicesPermission.remoteIndicesGroups()) {
+            for (IndicesPermission.Group indicesGroup : remoteIndicesGroup.indicesPermissionGroups()) {
+                indicesPrivileges.addAll(toIndicesPrivileges(indicesGroup));
+            }
+        }
+
+        // The role names matter for caching on the fulfilling cluster, when constructing the role cache key. A role descriptor can only
+        // store one name, whereas the role may have multiple names. To work around this, we will include the complete list of role names
+        // in the role descriptor metadata, and simply use the first name for the role descriptor as a place-holder.
+        final String roleDescriptorName = Arrays.stream(role.names()).iterator().next();
+        listener.onResponse(
+            new RoleDescriptorsIntersection(
+                List.of(
+                    Set.of(
+                        new RoleDescriptor(
+                            roleDescriptorName,
+                            null,
+                            indicesPrivileges.toArray(new RoleDescriptor.IndicesPrivileges[0]),
+                            null,
+                            null,
+                            null,
+                            // The fulfilling cluster should rely on this metadata field to construct the role key for caching
+                            Map.of("_role_names", role.names()),
+                            null
```

Comment:
I guess there may be more context on how this would work. But just based on this comment, I don't think we want to rely on the role names for caching. It feels backwards because the new RCS model wants to move away from relying on names from QC. It also does not really work by itself because different QCs can have same role names but different definitions. Or even the same QC can have different definitions of the same role name on different time. It's impossible for FC to reliably know when to invalidate the cache. 

I think we should compute the cache key based on the actual index privilege definition, similar to how API key roles are cached. This would require a slight change on how the header is serialised which I'll comment below.

Reply:
Caching like this makes a lot more sense for sure. We might still want to include the QC-side role names for auditing but for caching you're absolutely right.

Reply:
Aren't QC-side role names available in the authentication object already?

Reply:
This is tricky: @justincr-elastic and I discussed sanitizing the `authentication` instance we include in remote access headers, for the case when an API key is the subject on the querying cluster. In particular, we are considering omitting the role descriptors from the authentication metadata since those may contain DLS definitions that are sensitive (and not just DLS definitions for the querying cluster, also DLS definitions for _other_ fulfilling clusters). As such, for API key subjects the role names will not be available. 

We could of course modify the role descriptors included in the metadata in API keys, instead of omitting them altogether but it seems more straightforward to just leave them out. 

By including the role names in a metadata field here, the fulfilling cluster will have a complete list of role names that pertain to it and which can be cross-referenced between the FC and QC, if needed.

Reply:
Why do we want to preserve the role descriptor names of the QC side API key? Those are implementation artefacts and we intentionally do not want to rely on them. 

The current audit log has owner user's roles logged as `user.roles` when the authenticating subject is an API key. I personally consider it as a bug instead of something to replicate. Even if we want to follow this existing behaviour, on the FC side, the authenticating subject is the FC side API key (instead of QC side API key) and it can follow the existing pattern. Any audit log entries for QC side subject (API key or not) should be net new and I don't think we should use the role descriptors names in any of the new entries.